### PR TITLE
Added module for iron

### PR DESCRIPTION
**Pull request by rparree**
Initial commit to add support for iron.

Originally i added a PR to the Iron project (https://github.com/Iltotore/iron/pull/163), but it probably should be here.

I have been struggling with the build file and conflicts between cross compiling version. It might have to be double checked. Iron is a scala 3 project.

I have also not yet enabled support for JS.

### CODE REVIEW

**Comment by rparree** [comment]
I am having some difficulties troubleshooting the build. I already had to add two dummy subsequent git tags in order to make sure `previousStableVersion` was not `None`. When i run validate on the `IronJVM` project, it runs the test and works, but then breaks because `mimaPreviousClassfiles`, it is trying to obtain the previous published version of `ciris-iron`, which obviously is not yet available.

File: build.sbt [hunk 1]
```diff
@@ -182,6 +185,28 @@ lazy val refined = crossProject(JSPlatform, JVMPlatform, NativePlatform)
...
   .dependsOn(core)
 
+lazy val iron = crossProject(JVMPlatform)
+  .in(file("modules/iron"))
+  .settings(
+    moduleName := "ciris-iron",
+    name := moduleName.value,
+    dependencySettings ++ Seq(
+      libraryDependencies := Seq(
+        "io.github.iltotore" %% "iron" % ironVersion,
```
Comment (Iltotore) [inline build.sbt:195]: In order to support JS and Native, you should use `%%%` instead of `%%` for Iron dependency.

File: build.sbt [hunk 2]
```diff
@@ -182,6 +185,28 @@ lazy val refined = crossProject(JSPlatform, JVMPlatform, NativePlatform)
   .nativeSettings(sharedNativeSettings)
   .dependsOn(core)
 
+lazy val iron = crossProject(JVMPlatform)
```
Comment (Iltotore) [inline build.sbt:188]: To support JS and Native I guess you should add `JSPlatform` and `NativePlatform`.

You should also use `jsSettings` and `nativeSettings`.

File: build.sbt [hunk 3]
```diff
@@ -182,6 +185,28 @@ lazy val refined = crossProject(JSPlatform, JVMPlatform, NativePlatform)
...
+  .in(file("modules/iron"))
+  .settings(
+    moduleName := "ciris-iron",
+    name := moduleName.value,
+    dependencySettings ++ Seq(
+      libraryDependencies := Seq(
+        "io.github.iltotore" %% "iron" % ironVersion,
+        "org.scalameta" %% "munit" % "1.0.0-M8" % "test",
```
Comment (Iltotore) [inline build.sbt:196]: Shouldn't it be replaced with `mimaSettings`?

**Review by Iltotore — COMMENTED** [review]
Left some comments. Note I'm not familiar with Cirus neither SBT so I might be wrong.

File: build.sbt (hunk 2 above)
Comment (rparree) [reply build.sbt:188]: tx!, I'll look into JS and native once i have JVM working

**Comment by rparree** [comment]
locally i am stuck on: ` Error downloading is.cir:ciris_2.13:3.2.0+62-5d77a72c+20230804-1312-SNAPSHOT`

File: build.sbt (hunk 2 above)
Comment (rparree) [reply build.sbt]: Decided to already add js/native as it does not seem to interfere with the problem i'm faced with

**Comment by Iltotore** [comment]
@rparree did you solve your problem?

If not, do the other modules compile successfully?

File: build.sbt [hunk 4]
```diff
@@ -368,7 +401,7 @@ lazy val publishSettings =
 
 lazy val mimaSettings = Seq(
   mimaPreviousArtifacts := {
-    val unpublishedModules = Set[String]()
+    val unpublishedModules = Set[String]("ciris-iron")
```
Comment (Iltotore) [inline build.sbt:404]: Not sure what this code is supposed to do. As far as I can tell, this is the only difference with other modules. Maybe it is the cause of your dependency problem?

Comment (rparree) [reply build.sbt:404]: This is not part of my branch, i believe this was part of a recent commit by  [vlovgr](https://github.com/vlovgr)  https://github.com/vlovgr/ciris/pull/692/commits/55a3c98705e6efe2600246eb0f0738626245ce36

Comment (vlovgr) [reply build.sbt:404]: We add modules here when there's no published version yet, to exclude them from mima checks.

**Comment by vlovgr** [comment]
This shows up in the logs for the native version, not sure exactly what's causing this.
```
[error] Modules were resolved with conflicting cross-version suffixes in ProjectRef(uri("file:/home/runner/work/ciris/ciris/"), "ironNative"):
[error]    org.typelevel:cats-core_native0.4 _2.13, _3
[error]    io.github.cquiroz:cldr-api_native0.4 _2.13, _3
[error]    org.typelevel:cats-kernel_native0.4 _3, _2.13
[error]    org.typelevel:cats-effect-kernel_native0.4 _2.13, _3
[error]    io.github.cquiroz:scala-java-time_native0.4 _2.13, _3
[error]    io.github.cquiroz:scala-java-locales_native0.4 _2.13, _3

[error] java.lang.RuntimeException: Conflicting cross-version suffixes in: org.typelevel:cats-core_native0.4, io.github.cquiroz:cldr-api_native0.4, org.typelevel:cats-kernel_native0.4, org.typelevel:cats-effect-kernel_native0.4, io.github.cquiroz:scala-java-time_native0.4, io.github.cquiroz:scala-java-locales_native0.4
```

File: build.sbt (hunk 4 above)
Comment (rparree) [reply build.sbt:404]: OK that makes sense and why i was in a vicious circle.

**Comment by rparree** [comment]
Locally, I see "conflicting cross-version suffixes" on `cats-kernel` errors on all three: JS, Native and JVM (when running `sbt update`). I am pretty sure i had tackled those before.  I solved it then by using 

```
    scalaSettings ++ Seq(
      crossScalaVersions := Seq(scala3)
    ),
```

```
    scalaSettings ++ Seq(
      crossScalaVersions := Seq(scala3)
    ),
```

I revert back to that

**Comment by Iltotore** [comment]
You should not use Scala 2.x for Iron-related modules. It is Scala 3 only.

**Comment by Iltotore** [comment]
> But we can use it in a 2.13 project using CrossVersion.for2_13Use3

It shouldn't work because Iron relies heavily on Scala 3-only features like inline, S3 macros etc...

> AFAIK if we don't do this, we'll get two cats version on the dependency tree: the 2.13 from ciris and the 3.x one from iron.

Then I don't get the `.cross(CrossVersion.for2_13Use3)` 🤔. If the module uses Scala 3, why would we need this?

**Comment by rparree** [comment]
@vlovgr why is the build trying to download ciris version `ciris_2.13/3.2.0+65-e0f19883-SNAPSHOT`. It would never find something like that as `e0f19883` is my commit. Is this part of mima?

**Comment by vlovgr** [comment]
> @vlovgr why is the build trying to download ciris version `ciris_2.13/3.2.0+65-e0f19883-SNAPSHOT`. It would never find something like that as `e0f19883` is my commit. Is this part of mima?

It is part of MiMa, but I'm not sure why it happens. If we check `+mimaPreviousArtifacts` there's no mention of that version, but `+mimaPreviousArtifacts` yields the resolution errors you mention. The version which fails to resolve is the current version (e0f19883) and it only seems to fail on Scala 2.12.x (2.13.x and 3.x work without issues). When I manually remove support for Scala 2.12.x in the build, it works as expected. It might be that we're hitting an issue in MiMa related to Scala 2.12.x support? If we can't find a workaround, I'm happy to drop Scala 2.12.x support in a separate pull request.

**Comment by Iltotore** [comment]
So, what is the state of this PR? Should the support for 2.12.x be dropped?

**Comment by vlovgr** [comment]
> So, what is the state of this PR? Should the support for 2.12.x be dropped?

I've opened https://github.com/vlovgr/ciris/pull/701 as a proposal and have asked for feedback. I haven't had a chance to look into workarounds, but also haven't really got any ideas. Maybe one alternative could be to have this module in the iron repository, similar to other iron modules?

**Comment by Iltotore** [comment]
Actually, this was [already suggested](https://github.com/Iltotore/iron/pull/163) by @vlovgr but I suggested your repository instead, since the Refined module is also in yours.

I still prefer to have this module in this repository but if it happens to be too complicated to add, I can add it to Iron's repository.

I will try tomorrow to make this PR compile.

**Comment by Iltotore** [comment]
So I tried to fix this issue but I'm facing problems:
- the `previousArtifactVersion` key is always `None`, causing `NoSuchElementException`s when loading the build script
- my machine is not powerful enough to support the build. I always run out of memory and my CPU is overflowed.

Unfortunately, this last problem prevents me to further look for a fix.

Again, I hope someone will manage to make this PR compile but if it is too burdensome, I will merge the original PR in Iron's repository.

**Comment by Iltotore** [comment]
I finally decided to merge the original PR of @rparree in Iron's repository (https://github.com/Iltotore/iron/pull/163)

**Comment by vlovgr** [comment]
> I finally decided to merge the original PR of @rparree in Iron's repository ([Iltotore/iron#163](https://github.com/Iltotore/iron/pull/163))

That's great, thanks @Iltotore! 👍 Looks like we can close this pull request then.

**Pull request closed without merging**